interactive bots: Create GitHub issue bot.

fredfishgames opened the pull request:
This bot uses a github account owned by me using an alternate email address. It will post to the fredfishgames/zulip repository at the moment, which can be found [here](https://github.com/fredfishgames/zulip/issues) It will also push issues to the issues stream, followed by a tick or a cross depending on whether it was succesfully sent to github.

smarx commented:
*Automated message from Dropbox CLA bot*

@fredfishgames, it looks like you've already signed the Dropbox CLA. Thanks!

fredfishgames commented:
Ok, so it looks like I need to store the oauth token as an environment variable, as as soon as I upload the token to github, it is deleted, however, environment variables are machine specific. Therefore, should I simply encrypt the token so that github can't detect it, so that it can be used. The token is unable to do any real damage, as it only allows access to specific github functions.

fredfishgames commented:
Ok, so i think i'll upload the key to google drive, then download it in the script. This will mean we can update it in the future if the key gets reset

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+# See readme.md for instructions on running this code.
```
You have not added any Readme.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+        
+
+
+
```
I think this will fail linter test. You should add at most two blank lines.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+            '''
+
+      
+    
```
Two many blank lines.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+            PASSWORD = line
+        # Authentication for user filing issue DOES NOT WORK IF
+        # THE USER HAS READ/WRITE ACCESS TO THE REPOSITORY
+        USERNAME = 'zulip-issue-bot'
```
This should be added in the settings.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+
+        # The repository to add this issue to
+        REPO_OWNER = 'fredfishgames'
+        REPO_NAME = 'zulip'
```
Both of these should be added in settings.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+        data = urllib2.urlopen("https://drive.google.com/uc?export=download&id=0Bwo9Sux-T0A9QVFKV2FGb1RxbnM") #Downloads the file containing the token
+        for line in data: # files are iterable
+            print line
+            PASSWORD = line
```
Better to call this `oauth_token`.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+        #signs into github using the provided username and password
+        session = requests.session()
+        session.auth=(USERNAME, PASSWORD)
+        print('Logged in to github as %s' % (USERNAME))
```
Use `logging` instead of `print`.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+        #is_tag = False
+        #labels = ["from bot"]
+        
+        #current_tag = ""
```
Remove this commented out code.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+        
+        #Gets rid of the @issue in the issue title
+        issueTitle = message['content'].replace('@issue ', '')
+        issueTitle = message['content'].replace('@issue', '')
```
Change `issueTitle` to `issue_title` and use strip() `message['content'].replace('@issue', '').strip()`. I think you would probably want to limit length of title as well.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+         #           current_tag = ""
+         #       else:
+         #           current_tag += issueTitle[i]
+            
```
Remove this commented out code block.

umairwaheed commented:
@fredfishgames, awesome effort 👍. Just left some feedback, mostly dealing with the style. I think you would want just a single commit.

fredfishgames commented:
Thanks for the feedback!
I'll get working on these changes later today.

fredfishgames commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,149 @@
+        
+        print('sending issue')
+        #gets token from google drive
+        data = urllib2.urlopen("https://drive.google.com/uc?export=download&id=0Bwo9Sux-T0A9QVFKV2FGb1RxbnM") #Downloads the file containing the token
```
Oh

fredfishgames replied:
sorry, I was on the wrong account, so I couldn't see your messages

fredfishgames replied:
if it's not on github, how do I send my changes?

fredfishgames commented:
Oh, forgot to add in settings
Where is the settings file?

fredfishgames commented:
Ok, so: I've cleaned everything up, put all settings in ./zuliprc-local, and the key is now in dev-secrets.conf. However, both these files are only accessible on my remote instance. However, the code does work, but I am unable to update dev-secrets.conf, as it is not on github. So, hopefully you can see that my code works, but I if you would like proof, I can send some screenshots over. It might also work if you connect to my developer instance of zulip http://fredfishgames.zulipdev.org:9991/, and if you look at my github issue page https://github.com/fredfishgames/zulip/issues, you should be able to test it out using @issue. I will try to keep the script running.
I think I have fixed all the styling issues, and I have implemented all the features you have suggested.
Thanks,
Freddie

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,132 @@
+        
+        # signs into github using the provided username and password
+        session = requests.session()
+        session.auth=(USERNAME, oauth_token)
```
missing  spaces around `=`

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,132 @@
+        cut_string = False
+        cut_len = 0
+        for i in range(0, len(issue_title)):
+            if(i>70 and cut_string == False and issue_title[i] == ' '):
```
redundant perenthesis

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,132 @@
+            if(i>70 and cut_string == False and issue_title[i] == ' '):
+                cut_string = True
+            
+            if(cut_string == True):
```
just `if cut_string:`

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,132 @@
+                issue_content += issue_title[i]
+                cut_len += 1
+        if(cut_string == True):
+            issue_title = issue_title[:-(cut_len)]
```
why did you add parenthesis `:-cut_len`

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,132 @@
+
+
+  
+  
```
lot of empty lines at the end (please keep just one)

TomaszKolek commented:
The idea of the bot is that particular Zulip Admin can configure (or not) particular bot. So we don't need your config file - we should do it ourselves.

fredfishgames commented:
ok thanks! I've just been running lint test, and fixed certain styling errors, and I'll do all your suggestions now

fredfishgames commented:
Ok, just fixed and cleaned up everything and it **finally** passed the lint test!

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,123 @@
+        cut_string = False
+        cut_len = 0
+        for i in range(0, len(issue_title)):
+            if i > 70 and cut_string == False and issue_title[i] == ' ':
```
`not cut_string` or `cut_string is False`

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,123 @@
+        issue_content = ''
+        cut_string = False
+        cut_len = 0
+        for i in range(0, len(issue_title)):
```
I think you could change it to:
`for index, character in enumerate(issue_title)):`
then instead of `issue_title[i]` you could just write `character`

Sounds better?

fredfishgames replied:
Can I access the index of that character from the character variable, because I need to see if the current index is over 70

fredfishgames replied:
ah, sorry, I'm quite new to python (I haven't used it before starting the zulip tasks!), yeah, I'll implement that

TomaszKolek commented:
And additional comments ;-) Maybe something new for you ;-)

fredfishgames commented:
Added a comment for every line, hopefully thats enough :)

fredfishgames commented:
Oh, sorry, didn't see those earlier comments!

TomaszKolek commented on contrib_bots/lib/github.py:
```diff
@@ -0,0 +1,26 @@
+
+
+# Returns an authenticated session
+def auth():
```
Not sure if all of the comments are needed...

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+        config = ConfigParser.ConfigParser()
+        config.read([os.path.expanduser('~/.zuliprc-local')])
+
+        # The repo name to post to
```
I don't think comments are needed for those 3 lines?.
Those seem obvious.

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+
+        # Gets rid of the @issue in the issue title
+        issue_title = message['content'].replace('@issue', '').strip()
+        # Initialises the issue_content variable, the variable that the overflow goes into
```
If it's possible I would rather make comments inline. If line won't be too long of course! ;-(

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+            # and if there is a space (so that it does not cut in the middle of a row)
+            if index > 70 and not cut_string and character == ' ':
+                # If these are furfilled, the string should start to be cut
+                cut_string = True
```
can we simplify it to:
`cut_string = index > 70 and not cut_string and character == '  '`?

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+                cut_string = True
+            # Checks if the string should be cut
+            if cut_string:
+                # Adds the current letter of the issue title to the content, so it appears as a comment
```
only letters can be passed here?

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+        # After the loop has done, checks if the string has been cut at all
+        if cut_string:
+            # Takes away the cut bit of string that has been duplicated
+            issue_title = issue_title[:-cut_len]
```
Can we simplify it to:
`issue_title = "{}...".format(issue_title[:-cut_len])`?

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+            issue_title += '...'
+
+        # Creates the issue json, that is transmitted to the github api servers
+        issue = {'title': issue_title,
```
just formatting issue - move the title to new line and adjust the rest.

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+                }
+        # Add the issue to the repository
+        r = session.post(url, json.dumps(issue))
+        # Status code 201 means it has succeeded
```
This comment isn't needed I think.

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,135 @@
+        # This adds the sender's username, so that people can be recognised
+        new_content = original_content.replace('@issue',
+                                          'by %s:' % (original_sender,))
+        # gets the repo url
```
the comment is for sure not needed.

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+
+        # Gets rid of the @issue in the issue title
+        issue_title = message['content'].replace('@issue', '').strip()
+        issue_content = '' # Initialises the variable that the overflow goes into
```
https://www.python.org/dev/peps/pep-0008/#inline-comments

Please take a look at the instruction. There is a suggestion to have at least 2 space between code and `#`. Could you do this?

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+        # ups for own follow ups!
+        if message['display_recipient'] == 'issue':
+            return False
+        is_issue = (original_content.startswith('@issue'))
```
redundant parenthesis.

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+        issue_title = message['content'].replace('@issue', '').strip()
+        issue_content = '' # Initialises the variable that the overflow goes into
+        cut_string = False # cut_string - true if the string needs to be cut
+        cut_len = 0 # cut len - the length of the cut string
```
I don't see  reason to repeat name of variable after `#`

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+        # Creates the issue json, that is transmitted to the github api servers
+        issue = {
+                 'title': issue_title,
+                 'body': '{} **Sent by [{}](https://chat.zulip.org/#) from zulip**'
```
Isn't it impossible to have it in one line?

fredfishgames replied:
wouldn't it be too long, so it'll fail the lint test?

fredfishgames replied:
it might be ok

TomaszKolek commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+                  subject = message['sender_email'],
+                  # Adds a cross so that the user can see that it has failed, and provides a link to a
+                  # google search that can (hopefully) direct them to the error
+                  content = '{} :x: Code: [{}](https://www.google.co.uk/search?q=Github HTTP {} Error)'
```
maybe better to use `google.com`.

TomaszKolek replied:
Isn't it possible to have it in one line?

fredfishgames replied:
that line **is** too long

fredfishgames commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+              # This means that the issue has not been sent
+              # sends the message onto the 'issues' stream so it can be seen by zulip users
+              client.send_message(dict(
+                  type = 'stream',
```
should I apply that to every =?

fredfishgames commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,126 @@
+        cut_string = False # cut_string - true if the string needs to be cut
+        cut_len = 0 # cut len - the length of the cut string
+        # iterates for every letter in issue_title
+        for index, character in enumerate(issue_title):
```
well, issue_title would be a*40, if there are spaces between them, and so would to issue content

fredfishgames commented:
Hello, I've fixed all compatibility issues, and any errors that exist I've fixed, but it's not picking that up...

showell commented on contrib_bots/lib/github.py:
```diff
@@ -0,0 +1,22 @@
+    # Sets up the configParser to read the .conf file
+    config = six.moves.configparser.ConfigParser()
+    # Reads the config file
+    config.read([os.path.expanduser('~/.zuliprc-local')])
```
I know there's been some confusion around where we put the config info for github stuff, but I believe that putting it into `~/.zuliprc-local` is **not** the way to go.  One pitfall here is that a user might change their API key in Zulip and download a new zuliprc file, which will overwrite the github keys.

umairwaheed replied:
Yes, in the contrib-bot [Readme](https://github.com/zulip/zulip/blob/master/contrib_bots/lib/readme.md#third-party-configuration) it says to create a separate config file:
> 63 Do **not** put third party configuration information in your
 64 `.zuliprc` file.  Do not put third party configuration
 65 information anywhere in your Zulip directory.  Instead,
 66 create a separate configuration file for the third party's
 67 configuration in your home directory.

fredfishgames replied:
ok

umairwaheed commented:
>Hello, I've fixed all compatibility issues, and any errors that exist I've fixed, but it's not picking that up...

You mean the code is not updating on GitHub?

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,119 @@
+        else:
+              # This means that the issue has not been sent
+              # sends the message onto the 'issues' stream so it can be seen by zulip users
+                client.send_message(dict(
```
The indentation of this line does not match.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,119 @@
+
+        if r.status_code == 201:
+              # sends the message onto the 'issues' stream so it can be seen by zulip users
+                client.send_message(dict(
```
The indentation on this line does not match.

umairwaheed commented:
@fredfishgames, your code is beginning to look top-notch :). I think you just need to take care of the configuration file and some minor style issues and you are good to go.

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,119 @@
+                # Adds a cross so that the user can see that it has failed, and provides a link to a
+                # google search that can (hopefully) direct them to the error
+                content='{} :x: Code: [{}](https://www.google.com/search?q=Github HTTP {} Error)'
+                .format(new_content, r.status_code, r.status_code),
```
The actual text returned in the response could be also useful here. Also, I think his line should be indented.

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,119 @@
+        issue_content = ''
+        new_issue_title = ''
+        for part_of_title in issue_title.split():
+            if len(new_issue_title) < 70:
```
Could you make `70` it into a class constant so that it's better documented?

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,119 @@
+        # gets token from dev-secrets
+        # CONFIGURE TOKEN ON LOCAL MACHINE FOR NOW AT .zuliprc-local
+        config = six.moves.configparser.ConfigParser()
+        config.read([os.path.expanduser('~/.github-issue-bot/github-issue-bot.conf')])
```
The filename could be class constant so that it is easier locate if you want to change it in the future.

femnad commented on contrib_bots/lib/github.py:
```diff
@@ -0,0 +1,22 @@
+# The sole purpose of this is to authenticate, and it will
+# return the requests session that is properly authenticated
+import requests
+import six.moves.configparser, os
```
Python import convention suggest that modules are imported in different lines:
https://www.python.org/dev/peps/pep-0008/#imports

femnad replied:
Also having the imports in alphabetical order will look good too.

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,119 @@
+
+        new_content = original_content.replace('@issue', 'by {}:'.format(original_sender,))
+        # gets the repo url
+        url = 'https://api.github.com/repos/{}/{}/issues'.format(REPO_OWNER, REPO_NAME)
```
The url could be a class constant as well.

femnad commented on contrib_bots/lib/github.py:
```diff
@@ -0,0 +1,24 @@
+import requests
+import six.moves.configparser, os
+
+config_file = '~/.github-issue-bot/github-issue-bot.conf'
```
By convention Python constants should be in capital letters and words should be separated by underscores:
https://www.python.org/dev/peps/pep-0008/#constants

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,127 @@
+    as reporting it to github
+    '''
+
+    url = 'https://api.github.com/repos/{}/{}/issues'
```
By convention Python constants should be in capital letters and words should be separated by underscores:
https://www.python.org/dev/peps/pep-0008/#constants

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,127 @@
+standard_library.install_aliases()
+import urllib.request, urllib.error, urllib.parse
+import six.moves.configparser
+import os
```
Conventionally there is an expected grouping of Python imports: standard library imports, related third party imports and local application/library specific imports:

https://www.python.org/dev/peps/pep-0008/#imports

Also having them in alphabetical order will look good too.

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,127 @@
+    '''
+
+    url = 'https://api.github.com/repos/{}/{}/issues'
+
```
Empty line is not necessary here I think.

femnad commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,127 @@
+
+    def triage_message(self, message):
+        # return True if we want to (possibly) respond to this message
+
```
I think you could skip empty lines here since the comments provide good enough visual hints.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,122 @@
+        # Sends the HTTP post request
+        r = session.post(url_new, json.dumps(issue))
+
+        if r.status_code == 201:
```
Are you sure it's 201? Usually the success code is 200. To make it work with all 20x codes do this `if r.status_code / 100 == 2`; or if you are sure that they will only send 201 then just add a comment.

umairwaheed commented on contrib_bots/lib/github_issues.py:
```diff
@@ -0,0 +1,122 @@
+                subject=message['sender_email'],
+                # Adds a check mark so that the user can verify if it has been sent
+                content='{} :heavy_check_mark:'.format(new_content),
+            ))
```
May be you should just `return` from here, then you will not need to handle the error case in an `else`.

umairwaheed commented:
@fredfishgames, this is mostly done. Some minor feedback though.

reyha commented:
Hi @fredfishgames. Thanks for making all the necessary changes. Your code looks good to merge.

I would like to request @timabbott to see if this can now be merged. Thanks!

showell commented:
@reyha Have you had a chance to verify the code in your local dev environment?

fredfishgames commented:
Thanks!
I **really** appreciate all the help and feedback!
I didn't really know python before this task, let alone how to make web requests, and I have learned so much.
I'll be doing the github.py task next, so that it is easy for anyone to interface with the api without doing it directly.

showell commented:
Merged!  Thanks, @fredfishgames!
